In [7]:
from pyspark.sql import SparkSession, functions as F

# Create a spark session
spark = (
    SparkSession.builder.appName("BNPL External Data-set")
    .config("spark.sql.repl.eagerEval.enabled", True)
    .config("spark.sql.parquet.cacheMetadata", "true")
    .config("spark.driver.memory", "4g")
    .config("spark.sql.session.timeZone", "Etc/UTC")
    .getOrCreate()
)

In [95]:
# Store data in pyspark data frame:
population_df = spark.read.option("header", True).csv('../../../data/tables/population_data.csv')

In [62]:
# Inspect data frame:
population_df.limit(5)

+--------------------+------------+---------------+---------------+---------------+---------------+-------------+--------------+-------------+---------------+-----------------+--------------------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+-------------------------+---------------------------+---------+-------------------------------+--------------+--------------+------------------------+-------------------------+---------------------------+------------------------------+-------------------------+---------------------------+------------------------------+--------------+--------------+------------------------+-------------------------+---------------------------+------------------------------+-------------------------+---------------------------+------------------------------+--------------+--------------+------------------------+-------------------------+---------------------------+------------------------------+-------------------------+---------------------------+------------------------------+--------------+--------------+------------------------+-------------------------+---------------------------+------------------------------+-------------------------+---------------------------+------------------------------+--------------+--------------+------------------------+-------------------------+---------------------------+------------------------------+-------------------------+---------------------------+------------------------------+--------------------+
|                 FID|primaryindex|state_code_2016|state_name_2016|gccsa_code_2016|gccsa_name_2016|sa4_code_2016| sa4_name_2016|sa3_code_2016|  sa3_name_2016|sa2_maincode_2016|       sa2_name_2016|erp_2001|erp_2002|erp_2003|erp_2004|erp_2005|erp_2006|erp_2007|erp_2008|erp_2009|erp_2010|erp_2011|erp_2012|erp_2013|erp_2014|erp_2015|erp_2016|erp_2017|erp_2018|erp_2019|erp_2020|erp_2021|erp_change_number_2020_21|erp_change_per_cent_2020_21| area_km2|pop_density_2021_people_per_km2|births_2016_17|deaths_2016_17|natural_increase_2016_17|internal_arrivals_2016_17|internal_departures_2016_17|net_internal_migration_2016_17|overseas_arrivals_2016_17|overseas_departures_2016_17|net_overseas_migration_2016_17|births_2017_18|deaths_2017_18|natural_increase_2017_18|internal_arrivals_2017_18|internal_departures_2017_18|net_internal_migration_2017_18|overseas_arrivals_2017_18|overseas_departures_2017_18|net_overseas_migration_2017_18|births_2018_19|deaths_2018_19|natural_increase_2018_19|internal_arrivals_2018_19|internal_departures_2018_19|net_internal_migration_2018_19|overseas_arrivals_2018_19|overseas_departures_2018_19|net_overseas_migration_2018_19|births_2019_20|deaths_2019_20|natural_increase_2019_20|internal_arrivals_2019_20|internal_departures_2019_20|net_internal_migration_2019_20|overseas_arrivals_2019_20|overseas_departures_2019_20|net_overseas_migration_2019_20|births_2020_21|deaths_2020_21|natural_increase_2020_21|internal_arrivals_2020_21|internal_departures_2020_21|net_internal_migration_2020_21|overseas_arrivals_2020_21|overseas_departures_2020_21|net_overseas_migration_2020_21|                geom|
+--------------------+------------+---------------+---------------+---------------+---------------+-------------+--------------+-------------+---------------+-----------------+--------------------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+-------------------------+---------------------------+---------+-------------------------------+--------------+--------------+------------------------+-------------------------+---------------------------+------------------------------+-------------------------+---------------------------+------------------------------+--------------+--------------+-----

In [63]:
# Select subset important attributes, ignore the rest:
cols_to_keep = ['sa2_maincode_2016', 'sa2_name_2016', 'erp_2021']
population_df = population_df.select(*cols_to_keep)

In [64]:
# Inspect subset:
population_df.limit(3)

+-----------------+-------------+--------+
|sa2_maincode_2016|sa2_name_2016|erp_2021|
+-----------------+-------------+--------+
|        101021007|    Braidwood|    4360|
|        101021008|      Karabar|    8374|
|        101021009|   Queanbeyan|   11401|
+-----------------+-------------+--------+

In [65]:
# Check attribute types:
population_df.printSchema()

root
 |-- sa2_maincode_2016: string (nullable = true)
 |-- sa2_name_2016: string (nullable = true)
 |-- erp_2021: string (nullable = true)



In [66]:
# Assign correct attribute types from previous:
from pyspark.sql.types import IntegerType, LongType

population_df = population_df \
                .withColumn("sa2_maincode_2016", F.col("sa2_maincode_2016").cast(LongType())) \
                .withColumn("erp_2021", F.col('erp_2021').cast(IntegerType()))

population_df.printSchema()

root
 |-- sa2_maincode_2016: long (nullable = true)
 |-- sa2_name_2016: string (nullable = true)
 |-- erp_2021: integer (nullable = true)



In [67]:
# Rename columns:
population_df = population_df \
                .withColumnRenamed('sa2_name_2016', 'suburb') \
                .withColumnRenamed('erp_2021', 'estimated_region_population_2021')

In [96]:
# Function to be added to ETL:

from pyspark.sql.types import IntegerType, LongType

def population_preprocess(data):

    cols_to_keep = ['sa2_maincode_2016', 'sa2_name_2016', 'erp_2021']
    population_df = data.select(*cols_to_keep)

    population_df = population_df \
                    .withColumn("sa2_maincode_2016", F.col("sa2_maincode_2016").cast(LongType())) \
                    .withColumn("erp_2021", F.col('erp_2021').cast(IntegerType()))

    population_df = population_df \
                    .withColumnRenamed('sa2_name_2016', 'suburb') \
                    .withColumnRenamed('erp_2021', 'estimated_region_population_2021')
                    
    return population_df


In [97]:
# Testing if above function works ^^^:
test = population_preprocess(population_df)
test.limit(5)

+-----------------+--------------------+--------------------------------+
|sa2_maincode_2016|              suburb|estimated_region_population_2021|
+-----------------+--------------------+--------------------------------+
|        101021007|           Braidwood|                            4360|
|        101021008|             Karabar|                            8374|
|        101021009|          Queanbeyan|                           11401|
|        101031013|             Bombala|                            2411|
|        101041020|Bega-Eden Hinterland|                            8996|
+-----------------+--------------------+--------------------------------+

In [89]:
# Check number of entries:
test.count()

2292

According to:<br>
abs.gov.au/ausstats/abs@.nsf/lookup/by%20subject/1270.0.55.001~july%202016~main%20features~statistical%20area%20level%202%20(sa2)~10014 <br>

There are 2310 distinct SA2 codes, but the population data above only has 2292. Further discussion may be necessary.